src/sentry/auth/providers/saml2.py

```diff
@@ -103,241 +56,85 @@ def get_provider(organization_slug):
-            advanced_form = self.advanced_form_cls(initial=advanced_data)
-
-        return self.display_configure_view(
-            organization,
-            saml_form,
-            options_form,
-            attr_mapping_form,
-            advanced_form
-        )
-
-    def display_configure_view(self, organization, saml_form, options_form,
-                               attr_mapping_form, advanced_form):
-        raise NotImplementedError('Display Configure View not implemented!')
-
-
 class SAML2LoginView(AuthView):
     def dispatch(self, request, helper):
+        if 'SAMLResponse' in request.POST:
+            return helper.next_step()
+
         provider = helper.provider
-        if provider is None:
-            messages.add_message(request, messages.ERROR, ERR_NO_SAML_SSO)
-            return HttpResponseRedirect(request.path)
 
-        saml_config = provider.build_saml_config(helper.organization.slug)
-        auth = provider.build_auth(request, saml_config)
+        # If we're authenticating during the setup piepline the provider will
+        # not have been configured yet, build the config first from the state
+        if not provider.config:
+            provider.config = provider.build_config(helper.fetch_state())
+
+        saml_config = build_saml_config(provider.config, helper.organization.slug)
+        auth = build_auth(request, saml_config)
+
         return self.redirect(auth.login())
 
 
 class SAML2ACSView(AuthView):
     @method_decorator(csrf_exempt)
-    def dispatch(self, request, organization_slug):
-        if request.method != 'POST':
-            return HttpResponseNotAllowed(['POST'])
+    def dispatch(self, request, helper):
+        provider = helper.provider
 
-        provider = get_provider(organization_slug)
-        if provider is None:
-            messages.add_message(request, messages.ERROR, ERR_NO_SAML_SSO)
-            return HttpResponseRedirect('/')
+        # If we're authenticating during the setup piepline the provider will
+        # not have been configured yet, build the config first from the state
+        if not provider.config:
+            provider.config = provider.build_config(helper.fetch_state())
 
-        organization = Organization.objects.get(slug=organization_slug)
-        saml_config = provider.build_saml_config(organization_slug)
+        saml_config = build_saml_config(provider.config, helper.organization.slug)
 
-        auth = provider.build_auth(request, saml_config)
+        auth = build_auth(request, saml_config)
         auth.process_response()
-        errors = auth.get_errors()
-        if errors:
-            error_reason = auth.get_last_error_reason()
-            raise IdentityNotValid(error_reason)
 
-        attributes = auth.get_attributes()
-        nameid = auth.get_nameid()
+        # SSO response verification failed
+        if auth.get_errors():
+            return helper.error(ERR_SAML_FAILED.format(reason=auth.get_last_error_reason()))
```

Comment:
I don't think you can actually do this `.format()` here since `ERR_SAML_FAILED` isn't a string. It's a translation object. Usually these are used with just % formatting, but I'd just double check this.

Reply:
This does work, actually the `%` formatting wasn't working right: https://github.com/getsentry/sentry/pull/6173#discussion_r140884273